Node-based `--only-missing-outputs` for granular skipping

SajidAlamQB opened the pull request:
## Description
<!-- Why was this PR created? -->

This PR brings in the `--only-missing-outputs` CLI flag for `kedro run`.

The goal here is to skip a node if its outputs are already there and aren't needed, making re-runs faster.

**How it works:**

When `--only-missing-outputs` is used:

1.  We first figure out the "parent-child" relationships between nodes.
2.  Then, we go through the pipeline nodes in **reverse topological order** (from the end to start).
3.  For each node, we decide if it *must* run based on these main rules:
    *   **Node has no outputs?** This ensures side-effect nodes run if their part of the pipeline is active.
    *   **Node has its own missing persistent output?** Yes, it runs. This ensures any broken persistent node gets fixed, even for intermediate datasets.
    *   **Is an output of this node (ephemeral or persistent-and-missing) needed by a downstream "child" node that we've *already decided must run*?** Yes, then this "parent" node also runs to supply that child.


This MVP focuses on the *existence* of persistent datasets. More advanced change detection (like hashing to see if data or code is "stale") is a separate discussion for post-1.0.0 (see Spike: #4635).

**Development Notes:**

*   **CLI:** Uses the `--only-missing-outputs` flag.
*   **Session:** `KedroSession.run()` passes the `only_missing_outputs` flag to the runner.
*   **Runner Logic (`AbstractRunner`):**
    *   The main `run()` method calls `_filter_pipeline_for_missing_outputs()` if the flag is true.
    *   `_filter_pipeline_for_missing_outputs()`: handles the node-based filtering.
        *   Calls `_build_node_children_map()` to get dependencies.
        *   Calls `_determine_nodes_to_run()` which iterates in reverse topological order.
        *   `_determine_nodes_to_run()` uses `_should_node_run()` for each node.
    *   `_should_node_run()`: Goes through three rules mentioned above, calling helpers like:
        *   `_has_missing_persistent_outputs()`
        *   `_outputs_needed_by_children()`


## Testing Spaceflights

**Scenario: All Persistent Outputs Exist**

*   Run `kedro run` once to generate everything.
*   Run `kedro run --only-missing-outputs`.
*   **Expected:**
    *   Nodes producing only **persistent outputs** that all exist will be **skipped**. 

**Scenario: Delete `regressor.pickle`**

*   Run `kedro run` once.
*   Delete `data/06_models/regressor.pickle`.
*   Run `kedro run --only-missing-outputs`.
*   **Expected:**
    *   `train_model_node`: **Must run** (its persistent output `regressor` is missing).
    *   `split_data_node`: **Must run** (its ephemeral outputs `X_train`, `y_train` are needed by `train_model_node` which is running).
    *   `evaluate_model_node`: **Must run** (it has no outputs, so it's always a candidate; also, its input `regressor` is being recomputed).

**Scenario: Delete `model_input_table.parquet`**

*   Run `kedro run` once.
*   Delete `data/03_primary/model_input_table.parquet`.
*   Run `kedro run --only-missing-outputs`.
*   **Expected:**
    *   `create_model_input_table_node`: **Must run** (its persistent output `model_input_table` is missing).
    *   `split_data_node`: **Must run** 
    *   `evaluate_model_node`: **Must run** 

**Scenario: Dataset with Faulty Catalog Definition**

*   Temporarily edit `conf/base/catalog.yml` for `regressor`: change `type: pickle.PickleDataset` to `type: pandas.CSVDataset`.
*   Delete `data/06_models/regressor.pickle` if it exists.
*   Run `kedro run --only-missing-outputs`.
*   **Expected:** it **should fail** with an error.

**Scenario: Interaction with `--to-nodes split_data_node`**

*   Run `kedro run` once (all outputs exist, including `model_input_table.parquet`).
*   Run `kedro run --to-nodes split_data_node --only-missing-outputs`.
*   **Expected:**
    *  All nodes in this sub-pipeline are **SKIPPED**.


**Scenario: Running a specific pipeline (`--pipeline data_science`)**

*   Run `kedro run` once.
*   Delete `data/06_models/regressor.pickle`.
*   Run `kedro run --pipeline data_science --only-missing-outputs`.
*   **Expected:**
    *   All three nodes (`split_data_node`, `train_model_node`, `evaluate_model_node`) in the `data_science` pipeline will run.** `model_input_table.parquet` (input to the `data_science` pipeline) will be loaded as it exists.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes
- [x] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

rashidakanchwala commented:
I haven't looked at this code-wise. But I tested all the different scenarios on `demo-project` and it's working as expected :)

DimedS commented on kedro/runner/runner.py:
```diff
@@ -105,45 +111,193 @@ def run(
+
+            # Always run nodes with no outputs
+            if self._should_run_for_no_outputs(node):
+                run_this_node = True
```
Why do we always run nodes that have no outputs?

SajidAlamQB replied:
As Merel said, I don't have a strong preference myself. 

What does the team think is the more intuitive?

Example:

Should `evaluate_model_node` run if `regressor.pickle` was loaded from an existing file (and `train_model_node` was skipped)?

Or should it only run if `regressor.pickle` was actually re-generated by `train_model_node` in this specific `--only-missing-outputs` run?

Happy to adjust the logic.

DimedS reviewed (approved):
Great job, @SajidAlamQB! I really like the structure of the code - extending the run function with an additional pipeline filtering step is clean and intuitive. The algorithm itself is also now well designed, requiring just a single pipeline traversal with what appears to be linear complexity. I just have one small question.

ravi-kumar-pilla commented:
Awesome work @SajidAlamQB .... The code looks way better and the algorithm seems to cover corner cases. I tested the scenarios mentioned here and will need some more time for testing few more cases. But overall the code looks great apart from the minor [comment](https://github.com/kedro-org/kedro/pull/4859#discussion_r2160557890).

Also while testing (using spaceflights-pandas starter) for -

**Scenario: Delete model_input_table.parquet**

**Actual** - `train_model_node` , `preprocess_companies_node` and `preprocess_shuttles_node` are skipped which differs from expected. Why would these nodes run ? I think it should be `create_model_input_table_node` that must run as it rightly does

```bash
        INFO     Skipping 6 nodes with existing outputs:                          runner.py:213
                             compare_passenger_capacity_exp__822c8b63,                                     
                             compare_passenger_capacity_go__1e477111,                                      
                             create_confusion_matrix__2789b6a9, preprocess_companies_node,                 
                             preprocess_shuttles_node, train_model_node
```

**Scenario: Dataset with Faulty Catalog Definition**

- This is same as `kedro run`. Is the idea behind the test case to show the error handling is still the same ?

Thank you

SajidAlamQB commented:
>  think it should be `create_model_input_table_node` that must run as it rightly does

Thanks @ravi-kumar-pilla for reviewing and testing. You're correct this scenario is now fixed in description.

> **Scenario: Dataset with Faulty Catalog Definition**
> 
> * This is same as `kedro run`. Is the idea behind the test case to show the error handling is still the same ?
> 

Yes, to essentially show that we are "failing fast" and haven't changed fundamental error handling.

merelcht reviewed (commented):
I've done an initial review and left comments. I find the code quite hard to follow. I don't think the methods are broken up in the most logical way. 

I'm guessing you've used copilot on this code, because some of the terminology isn't very common in our codebase. Let's make sure to make code consistent with the existing base in terms of naming formats and terminology.

SajidAlamQB commented:
> I've done an initial review and left comments. I find the code quite hard to follow. I don't think the methods are broken up in the most logical way.

Hi @merelcht, I've updated the ordering of the runner code to flow more easily and make it more readable and broken down some of large nested code into smaller methods and adjusted language that has gone through an llm to align with our vocab this should be easier to read now. Let me know if you have any other thoughts.

merelcht reviewed (commented):
Thanks for addressing my comments @SajidAlamQB ! I've left a couple more, most important is about the placement of the methods.

merelcht commented on kedro/runner/runner.py:
```diff
@@ -690,3 +548,156 @@ def _find_initial_node_group(pipeline: Pipeline, nodes: Iterable[Node]) -> list[
+    return False
+
+
+def is_persistent_dataset_missing(
```
Any reason why this is a public method and all the others private?

merelcht reviewed (commented):
I think this is nearly there! The code is looking much cleaner and I find it easy to follow now 🙂 

One final suggestion I have is to add an e2e test for this feature in the `run.feature` file.
